### PR TITLE
Fixes #2677

### LINKED ISSUE #2677: Cargo Filters treat all potions as equal
## :round_pushpin: Description (REQUIRED)

Cargo filters do not know the difference between potions. Water bottle is the same for cargo as Mundane Potion, Potion of Healing, Potion of Regeneration, etc. the same problem for throwers and I assume that for lingering

## :bookmark_tabs: Steps to reproduce the Issue (REQUIRED)

1. Place the chest and fill it with several different potions of the same type (normal, throwing or lingering)
2. Connect the Cargo Node (Input) to the chest and enable the blacklist in it.
3. Place the second chest and connect the Advanced Cargo Node (Output) to it
4. Insert any potion of the selected type in the first chest into the whitelist filter.
Video: https://youtu.be/euOi2rSkB5M
## :bulb: Expected behavior (REQUIRED)

The cargo filter will distinguish potions.
## :scroll: Server Log

N/A

## :open_file_folder: /error-reports/ Folder

N/A

## :compass: Environment (REQUIRED)

image

### PR DESCRIPTION
**Pull request by svr333**
## Description
Fixes #2677 (all potions are treated as the same)
(Used #2831 as a basis)

## Proposed changes
Add potion meta check inside of SlimefunUtils#isItemSimilar()

## Related Issues (if applicable)
Resolves #2677

## Checklist

- [x] I have fully tested the proposed changes and promise that they will not break everything into chaos.
- [x] I have also tested the proposed changes in combination with various popular addons and can confirm my changes do not break them.
- [x] I followed the existing code standards and didn't mess up the formatting.
- [ ] I did my best to add documentation to any public classes or methods I added.
- [ ] I have added `Nonnull` and `Nullable` annotations to my methods to indicate their behaviour for null values
- [ ] I added sufficient Unit Tests to cover my code.

### CODE REVIEW

File: src/main/java/io/github/thebusybiscuit/slimefun4/utils/SlimefunUtils.java [hunk 1]
```diff
@@ -293,13 +294,20 @@ private static boolean equalsItemMeta(@Nonnull ItemMeta itemMeta, @Nonnull ItemM
...
             return false;
-        } else if (!checkLore) {
-            return true;
-        } else if (itemMeta.hasLore() && sfitemMeta.hasLore()) {
-            return equalsLore(itemMeta.getLore(), sfitemMeta.getLore());
-        } else {
-            return !itemMeta.hasLore() && !sfitemMeta.hasLore();
         }
+
+        if (checkLore) {
```
Comment (TheBusyBiscuit) [inline src/main/java/io/github/thebusybiscuit/slimefun4/utils/SlimefunUtils.java:299]:
```suggestion
        } else if (checkLore) {
```

**Review by TheBusyBiscuit — CHANGES_REQUESTED** [review]

File: src/main/java/io/github/thebusybiscuit/slimefun4/utils/SlimefunUtils.java [hunk 2]
```diff
@@ -293,13 +294,18 @@ private static boolean equalsItemMeta(@Nonnull ItemMeta itemMeta, @Nonnull ItemM
...
-        } else if (itemMeta.hasLore() && sfitemMeta.hasLore()) {
-            return equalsLore(itemMeta.getLore(), sfitemMeta.getLore());
-        } else {
-            return !itemMeta.hasLore() && !sfitemMeta.hasLore();
+        } else if (checkLore) {
+            if (itemMeta.hasLore() && sfitemMeta.hasLore() && !equalsLore(itemMeta.getLore(), sfitemMeta.getLore())
+                || itemMeta.hasLore() != sfitemMeta.hasLore()) {
+                return false;
+            }
+        }
```
Comment (TheBusyBiscuit) [inline src/main/java/io/github/thebusybiscuit/slimefun4/utils/SlimefunUtils.java:302]: This so overly complicated, I have no idea what that is supposed to say ._.

File: src/main/java/io/github/thebusybiscuit/slimefun4/utils/SlimefunUtils.java [hunk 3]
```diff
@@ -293,13 +294,20 @@ private static boolean equalsItemMeta(@Nonnull ItemMeta itemMeta, @Nonnull ItemM
...
-            return true;
-        } else if (itemMeta.hasLore() && sfitemMeta.hasLore()) {
-            return equalsLore(itemMeta.getLore(), sfitemMeta.getLore());
-        } else {
-            return !itemMeta.hasLore() && !sfitemMeta.hasLore();
+        } else if (checkLore) {
+            if (itemMeta.hasLore() && sfitemMeta.hasLore() && !equalsLore(itemMeta.getLore(), sfitemMeta.getLore())) {
+                return false;
+            }
+            else if (itemMeta.hasLore() != sfitemMeta.hasLore()) {
```
Comment (TheBusyBiscuit) [inline src/main/java/io/github/thebusybiscuit/slimefun4/utils/SlimefunUtils.java:301]: formatting

```suggestion
            } else if (itemMeta.hasLore() != sfitemMeta.hasLore()) {
```

**Review by TheBusyBiscuit — CHANGES_REQUESTED** [review]
Looks much more readable, do we have any unit tests for this method?
So we can ensure this wont break anything

**Comment by svr333** [comment]
I dont think we do

**Comment by TheBusyBiscuit** [comment]
:monkaS: concerning

File: src/main/java/io/github/thebusybiscuit/slimefun4/utils/SlimefunUtils.java [hunk 4]
```diff
@@ -293,13 +294,19 @@ private static boolean equalsItemMeta(@Nonnull ItemMeta itemMeta, @Nonnull ItemM
...
-        } else if (itemMeta.hasLore() && sfitemMeta.hasLore()) {
-            return equalsLore(itemMeta.getLore(), sfitemMeta.getLore());
-        } else {
-            return !itemMeta.hasLore() && !sfitemMeta.hasLore();
+        } else if (checkLore) {
+            if (itemMeta.hasLore() && sfitemMeta.hasLore() && !equalsLore(itemMeta.getLore(), sfitemMeta.getLore())) {
+                return false;
+            } else if (itemMeta.hasLore() != sfitemMeta.hasLore()) {
+                return false;
+            }
```
Comment (TheBusyBiscuit) [inline src/main/java/io/github/thebusybiscuit/slimefun4/utils/SlimefunUtils.java:302]: I think we should make variables here for `hasItemMetaLore` and `hasSfItemMetaLore`.

**Review by TheBusyBiscuit — CHANGES_REQUESTED** [review]

**Comment by svr333** [comment]
Do tests have to be written before this gets merged?

**Review by TheBusyBiscuit — APPROVED** [review]
LGTM

**Review by Sfiguz7 — APPROVED** [review]
LGTM too

**Pull request merged**